core/che-core-api-core/src/main/java/org/eclipse/che/api/core/util/Version.java

```diff
@@ -0,0 +1,224 @@
+        Matcher matcher = VERSION.matcher(version);
+        if (!matcher.find()) {
+            throw new IllegalArgumentException(format("Illegal version '%s'", version));
+        }
+
+        int hotFix = 0;
+        int milestone = 0;
+        int beta = 0;
+
+        String hotFixGroup = matcher.group(4);
+        if (!hotFixGroup.isEmpty()) {
+            hotFix = parseInt(hotFixGroup.substring(1));
+        }
+
+        String milestoneGroup = matcher.group(5);
+        if (!milestoneGroup.isEmpty()) {
+            milestone = parseInt(milestoneGroup.substring(MILESTONE_VERSION_SUFFIX.length()));
+        }
+
+        String betaGroup = matcher.group(6);
+        if (!betaGroup.isEmpty()) {
+            beta = parseInt(betaGroup.substring(BETA_VERSION_SUFFIX.length()));
+        }
+
+        return new Version(parseInt(matcher.group(1)),
+                           parseInt(matcher.group(2)),
+                           parseInt(matcher.group(3)),
+                           hotFix,
+                           milestone,
+                           beta,
+                           !matcher.group(7).isEmpty(),
+                           !matcher.group(8).isEmpty());
+    }
+
+    /**
+     * Checks if version suites for pattern.
+     * For example 3.1.0 version is suited for 3.1.* or 3.*.0
+     */
+    public boolean isSuitedFor(String versionRegex) {
+        Pattern pattern = compile(versionRegex);
+        return pattern.matcher(toString()).matches();
+    }
+
+    public boolean isSnapshot() {
+        return snapshot;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (!(o instanceof Version)) {
+            return false;
+        }
+
+        Version version = (Version)o;
+
+        if (bugFix != version.bugFix) {
+            return false;
+        }
+        if (hotFix != version.hotFix) {
+            return false;
+        }
+        if (major != version.major) {
+            return false;
+        }
+        if (milestone != version.milestone) {
+            return false;
+        }
+        if (beta != version.beta) {
+            return false;
+        }
+        if (minor != version.minor) {
+            return false;
+        }
+        if (rc != version.rc) {
+            return false;
+        }
+        if (snapshot != version.snapshot) {
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public int hashCode() {
+        int result = major;
+        result = 31 * result + minor;
+        result = 31 * result + bugFix;
+        result = 31 * result + hotFix;
+        result = 31 * result + milestone;
+        result = 31 * result + beta;
+        result = 31 * result + (snapshot ? 1 : 0);
+        result = 31 * result + (rc ? 1 : 0);
+        return result;
+    }
+
+    @Override
+    public int compareTo(Version o) {
+        if (major > o.major
```

Comment:
Really hard to understand the logic of this if block. Can we rework it to make it more clear? Or add a comment how versions should be compared?